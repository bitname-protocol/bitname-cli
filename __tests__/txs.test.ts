import {
    coin as Coin,
    keyring as KeyRing,
    tx as TX,
} from 'bcoin';

import {
    genRedeemScript,
    genLockTx,
    genUnlockTx,
} from '../lib/txs';
import {
    BadUserPublicKeyError,
    BadServicePublicKeyError,
    BadLockTransactionError,
} from '../lib/errors';

import * as fs from 'fs';
import * as path from 'path';

describe('tx generation', () => {
    it('generates a valid redeem script', () => {
        const serviceKey = Buffer.from('02a1633cafcc01ebfb6d78e39f687a1f0995c62fc95f51ead10a02ee0be551b5dc', 'hex');
        const userKey = Buffer.from('030589ee559348bd6a7325994f9c8eff12bd5d73cc683142bd0dd1a17abc99b0dc', 'hex');
        const script = genRedeemScript(userKey, serviceKey, 5);

        const expected = [
            'OP_IF',
            userKey.toString('hex'),
            'OP_CHECKSIG',
            'OP_ELSE',
            'OP_5',
            'OP_CHECKSEQUENCEVERIFY',
            'OP_DROP',
            serviceKey.toString('hex'),
            'OP_CHECKSIG',
            'OP_ENDIF',
        ];

        expect(script.toASM().split(' ')).toEqual(expected);
    });

    it('fails if user pubkey is invalid', () => {
        const serviceKey = Buffer.from('02a1633cafcc01ebfb6d78e39f687a1f0995c62fc95f51ead10a02ee0be551b5dc', 'hex');
        const userKey = Buffer.from('030000000000000000000000000000000000000000000000000000000000000000', 'hex');
        expect(() => {
            genRedeemScript(userKey, serviceKey, 5);
        }).toThrowError(BadUserPublicKeyError);
    });

    it('fails if service pubkey is invalid', () => {
        const serviceKey = Buffer.from('021111111111111111111111111111111111111111111111111111111111111111', 'hex');
        const userKey = Buffer.from('030589ee559348bd6a7325994f9c8eff12bd5d73cc683142bd0dd1a17abc99b0dc', 'hex');
        expect(() => {
            genRedeemScript(userKey, serviceKey, 1);
        }).toThrowError(BadServicePublicKeyError);
    });

    it('generates locking transactions', () => {
        const serviceKey = Buffer.from('02a1633cafcc01ebfb6d78e39f687a1f0995c62fc95f51ead10a02ee0be551b5dc', 'hex');

        const wif = 'cNJFgo1driFnPcBdBX8BrJrpxchBWXwXCvNH5SoSkdcF6JXXwHMm';
        const userRing = KeyRing.fromSecret(wif);

        const testCoin = {
            version: 1,
            height: -1,
            value: 100000000,
            hash: '453bbd02d4ef04be090ec79691e7f1749ac14141456c3394a513055fbc904bac',
        };

        const coins = [new Coin(testCoin)];

        const tx = genLockTx(coins, 'google', 1, 1, 10, userRing, serviceKey, 1);

        expect(tx.hash('hex')).toBe('631525eccd821de5518730aa4d543f53ed4b0d919c6f8d48b17f0cc3d4b9a70b');
    });

    it('generates user unlocking transaction', () => {
        const txDataPath = path.resolve(__dirname, 'data', 'valid_lock_tx.tx');
        const txData = fs.readFileSync(txDataPath, 'utf8').trim();

        const lockTX = TX.fromRaw(txData, 'hex');

        const wif = 'cUBuNVHb5HVpStD1XbHgafDH1QSRwcxUTJmueQLnyzwz1f5wmRZB';
        const userRing = KeyRing.fromSecret(wif);

        const tx = genUnlockTx(lockTX, 1, false, userRing, userRing.getPublicKey());

        expect(tx.hash('hex')).toBe('8ece940b4ef2692b6609e11f9933f624714f2256d7f003391c75df668e215f62');
    });

    it('generates service unlocking transaction', () => {
        const txDataPath = path.resolve(__dirname, 'data', 'valid_lock_tx.tx');
        const txData = fs.readFileSync(txDataPath, 'utf8').trim();

        const lockTX = TX.fromRaw(txData, 'hex');

        const wif = 'cUBuNVHb5HVpStD1XbHgafDH1QSRwcxUTJmueQLnyzwz1f5wmRZB';
        const userRing = KeyRing.fromSecret(wif);

        const tx = genUnlockTx(lockTX, 1, true, userRing, userRing.getPublicKey());

        expect(tx.hash('hex')).toBe('9c9cc53211f58c83ca4d85bc7a15a44fd133cb6f1b7e3f55f3d052798bd9598d');
    });

    it('errors on unlocking with incorrect user pubkey', () => {
        const txDataPath = path.resolve(__dirname, 'data', 'valid_lock_tx.tx');
        const txData = fs.readFileSync(txDataPath, 'utf8').trim();

        const lockTX = TX.fromRaw(txData, 'hex');

        const wif = 'cUTaW9nuwpwfuZLkgY98qnfdbzokta2BKxnQ43HyGf7jLEwe1Big';
        const userRing = KeyRing.fromSecret(wif);

        const serviceKey = Buffer.from('030589ee559348bd6a7325994f9c8eff12bd5d73cc683142bd0dd1a17abc99b0dc', 'hex');

        expect(() => {
            genUnlockTx(lockTX, 1, false, userRing, serviceKey);
        }).toThrow(BadLockTransactionError);
    });

    it('errors on unlocking with incorrect service pubkey', () => {
        const txDataPath = path.resolve(__dirname, 'data', 'valid_lock_tx.tx');
        const txData = fs.readFileSync(txDataPath, 'utf8').trim();

        const lockTX = TX.fromRaw(txData, 'hex');

        const wif = 'cUTaW9nuwpwfuZLkgY98qnfdbzokta2BKxnQ43HyGf7jLEwe1Big';
        const serviceRing = KeyRing.fromSecret(wif);

        const userKey = Buffer.from('030589ee559348bd6a7325994f9c8eff12bd5d73cc683142bd0dd1a17abc99b0dc', 'hex');

        expect(() => {
            genUnlockTx(lockTX, 1, true, serviceRing, userKey);
        }).toThrow(BadLockTransactionError);
    });

    it('errors on unlocking with incorrect privkey', () => {
        const txDataPath = path.resolve(__dirname, 'data', 'valid_lock_tx.tx');
        const txData = fs.readFileSync(txDataPath, 'utf8').trim();

        const lockTX = TX.fromRaw(txData, 'hex');

        const wif = 'cNJFgo1driFnPcBdBX8BrJrpxchBWXwXCvNH5SoSkdcF6JXXwHMm';
        const userRing = KeyRing.fromSecret(wif);

        expect(() => {
            genUnlockTx(lockTX, 1, false, userRing, userRing.getPublicKey());
        }).toThrow(BadLockTransactionError);
    });
});
