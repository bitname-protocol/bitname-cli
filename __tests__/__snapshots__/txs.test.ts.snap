// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tx generation generates locking transactions 1`] = `"be8ddfbdc7ed6c7b0ecf343592a300ced80ee2c14eeb36cf6a4c9fdbacd40009"`;

exports[`tx generation generates locking transactions until block 500000000 1`] = `"08af431b12f5894e88b8d164c7969a5f288a17daa3f4868d8f4470b8221553a5"`;

exports[`tx generation generates service unlocking transaction 1`] = `"04d0a072b5cb74d74259c93f1c14b656c999978ead92e2555fc5f6427c4ee567"`;

exports[`tx generation generates user unlocking transaction 1`] = `"130fdd54eef21b7718b6faea8a8a05a16ea6b315a7725bd70bf1ea7e7067d891"`;
